[docs] IDEFICS guide and task guides restructure

MKhalusova opened the pull request:
There are two now common approaches to solving tasks: fine-tuning specialized models, and prompting large models. While the fine-tuning approach is prominently featured in the docs, the latter one is not. At the same time, generation docs fall on the sidelines. 

This PR includes two parts: 
1. A guide to using IDEFICS for various image-text tasks such as image captioning, vqa, etc. This helps not only highlight the non-fine-tuning approach in the docs, but also showcase the new open source model.
2. TOC restructure. In this PR, I suggest a restructured table of contents where the task guides are now called "Guides for fine-tuning specialized models", and there's a new section on the same level called "Prompting and Generation Guides for Large Models". This new section currently contains the IDEFICS guide, and the "Text generation strategies" doc. In the future, we can add prompting guides, and updated text generation docs here. 
The new structure reflects two different approaches to tasks. 
Currently the zero-shot guides for specialized models are still under fine-tuning task guides, and we can discuss whether they should be moved under the new section or not.

MKhalusova committed:
initial commit for the IDEFICS task guide

MKhalusova committed:
conversational example

MKhalusova committed:
updated TOC

MKhalusova commented:
PR with images for the doc: https://huggingface.co/datasets/huggingface/documentation-images/discussions/175

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_26035). All of your documentation changes will be reflected on that endpoint.

MKhalusova committed:
fixed typos

MKhalusova commented:
@LysandreJik Please suggest who else should be added as a reviewer. Thanks!

stevhliu commented on docs/source/en/tasks/idefics.md:
```diff
@@ -0,0 +1,408 @@
+
+This model is exceptionally versatile and can be used for a wide range of image and multimodal tasks. Note, however, that 
+being a large model, it requires significant computational resources and infrastructure. It is up to you to decide whether 
+this approach suits your use case better than fine-tuning specialized models for each individual task. 
```
Might be helpful to do a brief pro and con of prompting versus fine-tuning to help users decide? For example, if you have a niche use-case maybe it is better to fine-tune instead of prompting.

MKhalusova replied:
Maybe, or we could have a whole conceptual guide on this topic. It feels like there's probably enough nuance for a whole doc.

stevhliu commented on docs/source/en/tasks/idefics.md:
````diff
@@ -0,0 +1,408 @@
+>>> model = IdeficsForVisionText2Text.from_pretrained(checkpoint, torch_dtype=torch.bfloat16, device_map="auto")
+```
+
+## Loading the quantized version of the model
````
I think this makes more sense as a third-level heading under the main topic of "Loading"

```suggestion
### Quantized model
```

stevhliu commented on docs/source/en/tasks/idefics.md:
```diff
@@ -0,0 +1,408 @@
+
+Photo by [Hendo Wang](https://unsplash.com/@hendoo). 
+
+IDEFICS accepts text and image prompts. However, to caption an image, you do not have to provide any text prompt to the 
```
```suggestion
IDEFICS accepts text and image prompts. However, to caption an image, you do not have to provide a text prompt to the 
```

stevhliu commented on docs/source/en/tasks/idefics.md:
```diff
@@ -0,0 +1,408 @@
+Photo by [Hendo Wang](https://unsplash.com/@hendoo). 
+
+IDEFICS accepts text and image prompts. However, to caption an image, you do not have to provide any text prompt to the 
+model, only the preprocessed input image. Without any text prompt, the model will start generating text from the 
```
```suggestion
model, only the preprocessed input image. Without a text prompt, the model will start generating text from the 
```

stevhliu commented on docs/source/en/tasks/idefics.md:
```diff
@@ -0,0 +1,408 @@
+model, only the preprocessed input image. Without any text prompt, the model will start generating text from the 
+BOS (beginning-of-sequence) token thus creating a caption.
+
+As image input to the model, you can use either an image object (PIL.Image) or a url from which the image can be retrieved.
```
```suggestion
As image input to the model, you can use either an image object (`PIL.Image`) or a url from which the image can be retrieved.
```

stevhliu commented on docs/source/en/tasks/idefics.md:
```diff
@@ -0,0 +1,408 @@
+<Tip>
+
+For longer outputs like this, you will greatly benefit from tweaking the text generation strategy. This can help 
+you significantly improve the quality of the generated output. Check out the [Text generation strategies doc](../generation_strategies.md) 
```
```suggestion
you significantly improve the quality of the generated output. Check out [Text generation strategies](../generation_strategies.md) 
```

stevhliu commented on docs/source/en/tasks/idefics.md:
```diff
@@ -0,0 +1,408 @@
+
+## Running inference in batch mode
+
+All of the earlier section illustrated using IDEFICS for a single example. In a very similar fashion, you can run inference 
```
```suggestion
All of the earlier sections illustrated IDEFICS for a single example. In a very similar fashion, you can run inference 
```

stevhliu commented on docs/source/en/tasks/idefics.md:
```diff
@@ -0,0 +1,408 @@
+
+## Visual question answering
+
+Visual Question Answering (VQA) is the task of answering open-ended questions based on an image. Similarly to image 
```
```suggestion
Visual Question Answering (VQA) is the task of answering open-ended questions based on an image. Similar to image 
```

stevhliu commented on docs/source/en/tasks/idefics.md:
```diff
@@ -0,0 +1,408 @@
+
+## Image-guided text generation
+
+For more creating applications, you can use image-guided text generation to generate text based on an image. This can be 
```
```suggestion
For more creative applications, you can use image-guided text generation to generate text based on an image. This can be 
```

stevhliu reviewed (commented):
Very nice IDEFICS guide!

Also pinging @patrickvonplaten and @gante since this kind of overlaps with the ongoing text generation work.

MKhalusova committed:
Apply suggestions from code review

Co-authored-by: Steven Liu <59462357+stevhliu@users.noreply.github.com>

patrickvonplaten commented:
Think this is a bit in conflict with this PR: https://github.com/huggingface/transformers/pull/25371

patrickvonplaten commented:
I'm not a huge fan of changing the more general "Task Guide" title to the more specific "Fine-tuning Guide" because I think many of the guides contain very valuable information that's just needed for inference.

E.g. let's say right now I want to know how to use a fine-tuned speech model from the Hub for speech transcription, I wouldn't necessarily look under "Fine-tuning Guides"
To be it would make more sense to either:
- Leave "Task Guide" as a title and integrate LLM prompting and Idefics under it
- Or completely split inference and training into two different areas of the docs.

patrickvonplaten commented:
Not sure what @gante thinks here

LysandreJik reviewed (approved):
Super cool guide! Only left nits. Would love to have Victor's review once he's back from leave.

Agree with Patrick that we don't need to change the "Task Guide" title

MKhalusova committed:
addressed feedback

MKhalusova commented:
re @patrickvonplaten 's concerns, I've reverted the TOC structure reorg, and placed the IDEFICS guide under Task Guides. 
I've also addressed @LysandreJik 's feedback.

patrickvonplaten reviewed (approved):
Thanks!

VictorSanh commented:
Amazing! thank you! looking at it today!

VictorSanh commented on docs/source/en/tasks/idefics.md:
```diff
@@ -0,0 +1,405 @@
+
+To illustrate the task, get an image to be captioned, e.g.:
+
+<div class="flex justify-center">
```
cuteness overload

VictorSanh commented on docs/source/en/tasks/idefics.md:
````diff
@@ -0,0 +1,405 @@
+A puppy in a flower bed
+```
+
+## Prompted image captioning
````
i was a bit surprised at first by the distinction between "image captioning" and "prompted image captioning" but i kind of like it. i feel like it's a gentle intro of prompts

VictorSanh commented on docs/source/en/tasks/idefics.md:
```diff
@@ -0,0 +1,405 @@
+
+While IDEFICS demonstrates great zero-shot results, your task may require a certain format of the caption, or come with 
+other restrictions or requirements that increase task's complexity. Few-shot prompting can be used to enable in-context learning.
+By providing examples in the prompt, you can steer the model to generate results that mimic given examples. 
```
```suggestion
By providing examples in the prompt, you can steer the model to generate results that mimic the format of given examples. 
```

VictorSanh reviewed (approved):
it looks awesome! thank you!
left a few minor comments

MKhalusova committed:
bad_words_ids

MKhalusova commented on docs/source/en/tasks/idefics.md:
```diff
@@ -0,0 +1,405 @@
+
+## Image classification
+
+IDEFICS is capable of classifying images into different categories without being explicitly trained on data containing 
+labeled examples from those specific categories. Given a list of categories and using its image and text understanding 
+capabilities, the model can infer which category the image likely belongs to. 
+
+Say, we have this image of a vegetable stand: 
```
I think we can mention it briefly.

MKhalusova committed:
Apply suggestions from code review

Co-authored-by: Victor SANH <victorsanh@gmail.com>

MKhalusova committed:
rank classification note

MKhalusova committed:
Merge remote-tracking branch 'origin/idefics-guide' into idefics-guide

ArthurZucker commented on docs/source/en/_toctree.yml:
```diff
@@ -88,6 +88,11 @@
     title: Generation
+  - isExpanded: false
+    sections:
+    - local: tasks/idefics
+      title: Image tasks with IDEFICS
+    title: Prompting
```
Seems to be the only task with a model name (`/tasks/idefics` , not very intuitive for me but since everyone seems okay with this no problem! 😉

MKhalusova replied:
Yes, it's in its own "Prompting section" which will also later have a guide on prompting LLMs for different tasks. So it's still about tasks, but with a different approach. prompting vs finetuning.

ArthurZucker commented on docs/source/en/tasks/idefics.md:
````diff
@@ -0,0 +1,424 @@
+Before you begin, make sure you have all the necessary libraries installed. 
+
+```bash
+pip install -q bitsandbytes sentencepiece accelerate transformers
````
Why not use ` tokenizers`  instead of sentencepiece 👿

MKhalusova replied:
It currently requires sentencepiece 🤷‍♀️

ArthurZucker replied:
That's my question haha why? The fast tokenizers should also work no?

MKhalusova replied:
It doesn't. Without explicitly installing sentencepiece I get a ValueError. 
```
ValueError: Couldn't instantiate the backend tokenizer from one of: 
(1) a `tokenizers` library serialization file, 
(2) a slow tokenizer instance to convert or 
(3) an equivalent slow tokenizer class to instantiate and convert. 
You need to have sentencepiece installed to convert a slow tokenizer to a fast one.
```

MKhalusova replied:
So we should leave the import in the doc for now.

ArthurZucker commented on docs/source/en/tasks/idefics.md:
```diff
@@ -0,0 +1,424 @@
+...     bnb_4bit_compute_dtype=torch.float16,
+... )
+
+>>> processor = AutoProcessor.from_pretrained(checkpoint)
```
If `tokenizers`  is installed will use the fast by default here

ArthurZucker commented on docs/source/en/tasks/idefics.md:
````diff
@@ -0,0 +1,424 @@
+```
+
+<Tip>
+It is a good idea to include the `bad_words_ids` in the call to `generate` to avoid errors arising when increasing 
+the `max_new_tokens`: the model will want to generate a new `<image>` or `<fake_token_around_image>` token when there 
+is no image being generated by the model.
````
But this also induced an overhead related to the ` logit_processor` . If the default `generation_config` uses beamsearch it' s fine I guess it won' t make a huge difference

MKhalusova replied:
This is a recommendation from @VictorSanh

ArthurZucker commented on docs/source/en/tasks/idefics.md:
```diff
@@ -0,0 +1,424 @@
+... ]
+
+>>> inputs = processor(prompt, return_tensors="pt").to("cuda")
+>>> bad_words_ids = processor.tokenizer(["<image>", "<fake_token_around_image>"], add_special_tokens=False).input_ids
```
we can hint that badword ids can also be stored in the generation config to not have to always define them no?

MKhalusova replied:
I'll add a note

ArthurZucker reviewed (approved):
Thanks 🤗

MKhalusova committed:
feedback addressed
